src/runtime/security/kinit_context.cpp

```diff
@@ -0,0 +1,306 @@
+
+    // KRB5_LNAME_NOTRANS means no translation available for requested principal
+    if (err == KRB5_LNAME_NOTRANS) {
+        if (_principal->length > 0) {
+            int cnt = 0;
+            while (cnt < _principal->length) {
+                std::string tname;
+                tname.assign((const char *)_principal->data[cnt].data,
+                             _principal->data[cnt].length);
+                if (!_user_name.empty()) {
+                    _user_name += '/';
+                }
+                _user_name += tname;
+                cnt++;
+            }
+            return error_s::ok();
+        }
+        return error_s::make(ERR_KRB5_INTERNAL, "parse username from principal failed");
+    }
+
+    // KRB5_CONFIG_NOTENUFSPACE means BUF_LEN is not enough
+    if (err == KRB5_CONFIG_NOTENUFSPACE) {
+        return error_s::make(ERR_KRB5_INTERNAL, fmt::format("username is larger than {}", BUF_LEN));
+    }
+    KRB5_RETURN_NOT_OK(err, "krb5 parse aname to localname failed");
+
+    if (strlen(buf) <= 0) {
+        return error_s::make(ERR_KRB5_INTERNAL, "empty username");
+    }
+
+    _user_name.assign((const char *)buf);
+    return error_s::ok();
+}
+
+error_s kinit_context_impl::get_credentials()
+{
+    krb5_creds creds;
+    error_s err = error_s::ok();
+
+    // get initial credentials using a key table
+    // Notice: the contents of a krb5_creds structure need to be freed by ourselves
+    err = wrap_krb5_err(krb5_get_init_creds_keytab(_krb5_context,
+                                                   &creds,
+                                                   _principal,
+                                                   _keytab,
+                                                   0 /*valid from now*/,
+                                                   nullptr /*empty TKT service name*/,
+                                                   _opt),
+                        "get_init_cred");
+    if (!err.is_ok()) {
+        dwarn_f("get credentials of {} from KDC failed, reason({})",
+                FLAGS_krb5_principal,
+                err.description());
+        return err;
+    }
+    auto cleanup = dsn::defer([&]() { krb5_free_cred_contents(_krb5_context, &creds); });
+
+    // store credentials into _ccache.
+    err = wrap_krb5_err(krb5_cc_store_cred(_krb5_context, _ccache, &creds), "store_cred");
+    if (!err.is_ok()) {
+        dwarn_f("store credentials of {} to cache failed, err({})",
+                FLAGS_krb5_principal,
+                err.description());
+        return err;
+    }
+
+    _cred_expire_timestamp = creds.times.endtime;
+    ddebug_f("get credentials of {} from KDC ok, expires at {}",
+             FLAGS_krb5_principal,
+             utils::time_s_to_date_time(_cred_expire_timestamp));
+    return err;
+}
+
+void kinit_context_impl::schedule_renew_credentials()
+{
+    // reserve 300 seconds for renew
```

Comment:
Why use 300 seconds as the interval for renew?

Reply:
Let me have a look. I didn't notice it before :(